### ISSUE TITLE
Benchmark and test different indexes for hashrates table

### ISSUE THREAD

**Issue by tdroxler** [alephium/explorer-backend#158]
As discussed here: https://github.com/alephium/explorer-backend/pull/153#discussion_r816664996

**Comment by simerplaha** [comment]
In the comment I said 

> Just tried looking if I had benchmarks documented anywhere for adding partial index (main_chain = true) to block_headers table (https://github.com/alephium/explorer-backend/pull/131) but I think I lost it.

These benchmarks were documented [here](https://github.com/alephium/explorer-backend/pull/116#issuecomment-1019985980). They ran directly on the `block_headers` table and showed that performance was fluctuating.

This time I chose to run the benchmark on a fresh table with only the index being benchmarked created, aiming for consistent benchmark results. 

# Benchmark of partial VS full indexes

- `user_type` column is used to benchmark `interval_type` in table `hashrates`
- `active` column is used to benchmark `main_chain` in table `block_headers`

```sql
CREATE TABLE users
(
    id         int not null,
    first_name varchar(255) not null,
    last_name  varchar(255) not null,
    active     boolean not null,
    user_type  int not null
);
```

- `NoIndex` - Query without any index
- `PartialIndex` - Query with partial index (eg: `main_chain = true` or `interval_type = 0`)
- `FullIndex` - Basic full index on the field

The benchmarks are on both fields of type `active (Boolean - true or false)` and `user_type (Integer - 0 or 1)`.

![image](https://user-images.githubusercontent.com/1773953/159843047-28d3d07f-0378-4a6f-9c65-4da3b67a8562.png)

# Result

I've run these benchmark multiple times and the outcome is the same. There is no strong proof that partial index perform better when data value range is small (2 in this case). Being B-tree indexes I would assume the root node of tree would already split the tree into two halves for `boolean` and `integer (0 or 1)`  therefore the performance difference being unnoticeable. 

We do save nearly 50% on storage using `partial index` when the data is 50-50.

| partial_index | full_index |
|---------------|------------|
| 3168 kB       | 6328 kB    |

I could run more benchmarks with more randomness but I didn't want to overdo this.

**Comment by tdroxler** [comment]
thx again for those amazing benchmarks. So not much difference of performance with full or partial index, but still better to use to save some spaces right?

**Comment by simerplaha** [comment]
Yep I guess so :)

**Comment by simerplaha** [comment]
Yep I guess so :) 

Theoretically it makes sense that partial index would be faster. We haven't found a practical use-case for it yet other than space savings.

**Comment by polarker** [comment]
Thanks for the great benchmarks ❤️ 

Just be curious: is the ID unique in your benchmark? If there are duplicated id with different `active` and `user_type`, I would expect some performance hit.

**Comment by simerplaha** [comment]
In this case duplicate ids wouldn't cause any performance difference because the `id` column is not being used in the benchmark query.

But they are unique.

Both these queries return 1_000_000;
```sql
select count(*) as row_count from users_fullactive;
select count(distinct id) as distinct_id_count from users_fullactive;
```

**Comment by simerplaha** [comment]
The benchmarks are documented. So this issue is resolved. 

The actual solution to improve indexing to save storage space will be implemented via #191 as mentioned by the [comment](https://github.com/alephium/explorer-backend/issues/191#issuecomment-1091341595), after other priority tasks are implemented.
